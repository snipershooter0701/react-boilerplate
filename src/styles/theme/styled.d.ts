import 'styled-components';
import { defaultTheme } from './themes';

type Theme = typeof defaultTheme;

/* This is the suggested way of declaring theme types */
declare module 'styled-components' {
  export interface DefaultTheme extends Theme {}
}
