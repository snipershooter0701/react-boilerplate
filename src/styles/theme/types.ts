import { themes } from './themes';

export type ThemeKeyType = keyof typeof themes;
export interface ThemeState {
  selected: ThemeKeyType;
}
