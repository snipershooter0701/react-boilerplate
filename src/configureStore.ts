/**
 * Create the store with dynamic reducers
 */

import { applyMiddleware, createStore, compose, StoreEnhancer } from 'redux';
import { routerMiddleware } from 'connected-react-router';
import { createInjectorsEnhancer, forceReducerReload } from 'redux-injectors';
import createSagaMiddleware from 'redux-saga';
import { History } from 'history';
import { composeWithDevTools } from 'redux-devtools-extension';

import createReducer from './reducers';
import { ApplicationRootState, InjectedStore } from 'types';

export default function configureStore(
  initialState: ApplicationRootState | {} = {},
  history: History,
) {
  const reduxSagaMonitorOptions = {};
  const sagaMiddleware = createSagaMiddleware(reduxSagaMonitorOptions);
  const { run: runSaga } = sagaMiddleware;

  // Create the store with two middlewares
  // 1. sagaMiddleware: Makes redux-sagas work
  // 2. routerMiddleware: Syncs the location/URL path to the state
  const middlewares = [sagaMiddleware, routerMiddleware(history)];

  const enhancers = [
    applyMiddleware(...middlewares),
    createInjectorsEnhancer({
      createReducer,
      runSaga,
    }),
  ];

  let enhancer: StoreEnhancer;
  // If Redux Dev Tools and Saga Dev Tools Extensions are installed, enable them
  /* istanbul ignore next */
  if (process.env.NODE_ENV !== 'production' && typeof window === 'object') {
    enhancer = composeWithDevTools(...enhancers);
    // NOTE: Uncomment the code below to restore support for Redux Saga
    // Dev Tools once it supports redux-saga version 1.x.x
    // if (window.__SAGA_MONITOR_EXTENSION__)
    //   reduxSagaMonitorOptions = {
    //     sagaMonitor: window.__SAGA_MONITOR_EXTENSION__,
    //   };
  } else {
    enhancer = compose(...enhancers);
  }

  const store = createStore(
    createReducer(),
    initialState,
    enhancer,
  ) as InjectedStore;

  // Extensions
  store.runSaga = sagaMiddleware.run;
  store.injectedReducers = {}; // Reducer registry
  store.injectedSagas = {}; // Saga registry

  // Make reducers hot reloadable, see http://mxs.is/googmo
  /* istanbul ignore next */
  if (module.hot) {
    module.hot.accept('./reducers', () => {
      forceReducerReload(store);
    });
  }

  return store;
}
