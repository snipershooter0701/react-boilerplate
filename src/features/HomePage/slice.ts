/*
 * ReposManager Slice
 *
 * Here we define:
 * - The shape of our container's slice of Redux store,
 * - All the actions which can be triggered for this slice, including their effects on the store.
 *
 * Note that, while we are using dot notation in our reducer, we are not actually mutating the state
 * manually. Under the hood, we use immer to apply these updates to a new copy of the state.
 * Please see https://immerjs.github.io/immer/docs/introduction for more information.
 *
 */

import { PayloadAction } from '@reduxjs/toolkit';
import { createSlice } from 'utils/@reduxjs/toolkit';
import { HomePageState } from './types';
import { Repo } from 'types/Repo';

// The initial state of the homepage container
export const initialState: HomePageState = {
  username: '',
  repositories: [],
  loading: false,
  error: false,
};

const homepageSlice = createSlice({
  name: 'homepage',
  initialState,
  reducers: {
    changeUsername(state, action: PayloadAction<string>) {
      // Delete prefixed '@' from the github username
      const username = action.payload.replace(/@/gi, '');
      state.username = username;
    },
    loadRepos(state) {
      state.loading = true;
      state.error = false;
      state.repositories = [];
    },
    reposLoaded(state, action: PayloadAction<Repo[]>) {
      const repos = action.payload;
      state.repositories = repos;
      state.loading = false;
    },
    repoLoadingError(state) {
      state.error = true;
      state.loading = false;
    },
  },
});

export const { actions, reducer, name: sliceKey } = homepageSlice;
