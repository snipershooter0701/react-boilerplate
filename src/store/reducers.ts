/**
 * Combine all reducers in this file and export the combined reducers.
 */

import { combineReducers } from '@reduxjs/toolkit';
import { connectRouter } from 'connected-react-router';

import history from 'utils/history';
import { InjectedReducersType } from 'utils/types/injector-typings';

/**
 * Merges the main reducer with the router state and dynamically injected reducers
 */
export default function createReducer(
  injectedReducers: InjectedReducersType = {},
) {
  const rootReducer = combineReducers({
    ...injectedReducers,
    router: connectRouter(history),
  });

  return rootReducer as any; // TODO: any
}
