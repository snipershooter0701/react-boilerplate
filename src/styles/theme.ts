export const theme = {
  primary: '',
  componentBackground: '',
  componentBackgroundSecondary: '',
};
