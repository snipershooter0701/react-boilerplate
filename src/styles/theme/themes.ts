const defaultTheme = {
  primary: '',
  componentBackground: '',
  componentBackgroundSecondary: '',
};

const darkTheme: Theme = {
  primary: '',
  componentBackground: '',
  componentBackgroundSecondary: '',
};

export type Theme = typeof defaultTheme;

export const themes = {
  default: defaultTheme,
  dark: darkTheme,
};
