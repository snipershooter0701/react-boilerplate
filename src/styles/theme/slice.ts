import { PayloadAction, createSelector, createSlice } from '@reduxjs/toolkit';
import { ThemeState, ThemeKeyType } from './types';
import { themes } from './themes';
import { getThemeFromStorage, isSysyemDark } from './utils';

export const initialState: ThemeState = {
  selected: getThemeFromStorage() || 'default',
};

const themeSlice = createSlice({
  name: 'theme',
  initialState,
  reducers: {
    changeTheme(state, action: PayloadAction<ThemeKeyType>) {
      state.selected = action.payload;
    },
  },
});

export const selectTheme = createSelector(
  [(state: { theme: ThemeState }) => state.theme || initialState],
  theme => {
    if (theme.selected === 'system') {
      return isSysyemDark ? themes.dark : themes.default;
    }
    return themes[theme.selected];
  },
);

export const selectThemeKey = createSelector(
  [(state: { theme: ThemeState }) => state.theme || initialState],
  theme => theme.selected,
);

export const { changeTheme } = themeSlice.actions;
export const reducer = themeSlice.reducer;
export const themeSliceKey = themeSlice.name;
