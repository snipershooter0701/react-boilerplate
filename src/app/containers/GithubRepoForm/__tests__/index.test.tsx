import React from 'react';

import {
  Store,
  ActionCreatorWithoutPayload,
  PayloadAction,
  ActionCreator,
  Action,
} from '@reduxjs/toolkit';
import { render, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import { ThemeProvider } from 'styles/theme/ThemeProvider';
import { HelmetProvider } from 'react-helmet-async';
import { GithubRepoForm } from '..';
import { configureAppStore } from 'store/configureStore';
import { actions, initialState } from '../slice';

const renderGithubRepoForm = (store: Store) =>
  render(
    <Provider store={store}>
      <ThemeProvider>
        <HelmetProvider>
          <GithubRepoForm />
        </HelmetProvider>
      </ThemeProvider>
    </Provider>,
  );

const setUsernameInStoreAndUnmount = (store: Store, username: string) => {
  // Render, update the state and unmount
  const { container, unmount } = renderGithubRepoForm(store);
  const input = container.querySelector('input');
  fireEvent.change(input!, { target: { value: username } });
  unmount();
};

describe('<GithubRepoForm />', () => {
  let store: ReturnType<typeof configureAppStore>;
  let mockLoadRepos: jest.Mock<Action<string>>;

  beforeAll(() => {
    mockLoadRepos = jest.fn(() => ({
      type: '',
    }));
    actions.loadRepos = mockLoadRepos as any;
    actions.loadRepos.type = 'different_type_so_that_saga_wont_trigger';
  });

  beforeEach(() => {
    store = configureAppStore();
    mockLoadRepos.mockClear();
  });

  it('should match the snapshot', () => {
    const form = renderGithubRepoForm(store);
    expect(form.container.firstChild).toMatchSnapshot();
  });

  it("should fetch repos on mount if username isn't empty", () => {
    renderGithubRepoForm(store);
    expect(initialState.username.length).toBeGreaterThan(0);
    expect(mockLoadRepos).toHaveBeenCalledTimes(1);
  });

  it("shouldn't fetch repos on mount if username is empty", () => {
    setUsernameInStoreAndUnmount(store, '');
    mockLoadRepos.mockClear();

    // Now render again to trigger useEffect
    renderGithubRepoForm(store);
    expect(mockLoadRepos).not.toHaveBeenCalled();
  });
});
