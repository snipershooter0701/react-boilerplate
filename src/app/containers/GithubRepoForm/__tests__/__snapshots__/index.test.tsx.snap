// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GithubRepoForm /> should match the snapshot 1`] = `
.c3 {
  text-transform: uppercase;
  font-weight: normal;
  margin: 0;
  padding: 0;
  color: rgba(58,52,51,0.7);
  font-size: 0.75rem;
}

.c6 {
  border: 2px solid rgba(58,52,51,0.12);
  border-radius: 4px;
  color: rgba(58,52,51,1);
  display: block;
  width: 100%;
  font-size: 1rem;
  font-weight: 500;
  line-height: 1.5;
  color: rgba(58,52,51,1);
  background-color: rgba(255,255,255,1);
  -webkit-transition: border-color 0.15s ease-in-out,box-shadow 0.15s ease-in-out;
  transition: border-color 0.15s ease-in-out,box-shadow 0.15s ease-in-out;
  outline: none;
  height: 2.75rem;
  padding: 0 0.75rem;
}

.c6::-webkit-input-placeholder {
  font-weight: 300;
  color: rgba(58,52,51,0.7);
}

.c6::-moz-placeholder {
  font-weight: 300;
  color: rgba(58,52,51,0.7);
}

.c6:-ms-input-placeholder {
  font-weight: 300;
  color: rgba(58,52,51,0.7);
}

.c6::placeholder {
  font-weight: 300;
  color: rgba(58,52,51,0.7);
}

.c6:focus {
  border-color: rgba(215,113,88,1);
  box-shadow: 0 0 0 3px rgba(215,113,88,0.2);
}

.c0 .TextButton-b1pz3m-0 {
  margin: 16px 0;
  font-size: 0.875rem;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c4 .c5 {
  width: 33.333333333333336%;
  margin-right: 0.5rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-bottom: 1rem;
}

.c1 .c2 {
  margin-bottom: 0.25rem;
  margin-left: 0.125rem;
}

<div
  class="c0"
>
  <form
    class="c1"
  >
    <label
      class="c2 c3"
    >
      Github Username
    </label>
    <div
      class="c4"
    >
      <input
        class="c5 c6"
        placeholder="Type any Github username"
        type="text"
        value="react-boilerplate"
      />
    </div>
  </form>
</div>
`;
